src/treesitter.rs

```diff
@@ -21,3 +30,305 @@ impl From<&'_ Node<'_>> for Match {
+    type I = RopeOrSliceTextProviderIterator<'a>;
+
+    fn text(&mut self, node: Node) -> Self::I {
+        match self {
+            RopeOrSlice::Slice(slice) => {
+                RopeOrSliceTextProviderIterator::Slice(iter::once(&slice[node.byte_range()]))
+            }
+            RopeOrSlice::Rope(rope) => {
+                let rope_slice = rope.byte_slice(node.byte_range());
+                RopeOrSliceTextProviderIterator::Rope(RopeOrSliceRopeTextProviderIterator::new(
+                    rope_slice,
+                    |rope_slice| rope_slice.chunks(),
+                ))
+            }
+        }
+    }
+}
+
+impl<'a> Parseable for RopeOrSlice<'a> {
+    fn parse(&self, parser: &mut Parser, old_tree: Option<&Tree>) -> Option<Tree> {
+        match self {
+            Self::Slice(slice) => slice.parse(parser, old_tree),
+            Self::Rope(rope) => rope.parse(parser, old_tree),
+        }
+    }
+}
+
+impl<'a> Parseable for &'a RopeOrSlice<'a> {
+    fn parse(&self, parser: &mut Parser, old_tree: Option<&Tree>) -> Option<Tree> {
+        match self {
+            RopeOrSlice::Slice(slice) => slice.parse(parser, old_tree),
+            RopeOrSlice::Rope(rope) => rope.parse(parser, old_tree),
+        }
+    }
+}
+
+impl<'a> From<&'a [u8]> for RopeOrSlice<'a> {
+    fn from(value: &'a [u8]) -> Self {
+        Self::Slice(value)
+    }
+}
+
+impl<'a> From<&'a Rope> for RopeOrSlice<'a> {
+    fn from(value: &'a Rope) -> Self {
+        Self::Rope(value)
+    }
+}
+
+pub enum RopeOrSliceTextProviderIterator<'a> {
+    Slice(iter::Once<&'a [u8]>),
+    Rope(RopeOrSliceRopeTextProviderIterator<'a>),
+}
+
+impl<'a> Iterator for RopeOrSliceTextProviderIterator<'a> {
+    type Item = &'a [u8];
+
+    fn next(&mut self) -> Option<Self::Item> {
+        match self {
+            Self::Slice(slice_iterator) => slice_iterator.next(),
+            Self::Rope(rope_iterator) => rope_iterator.next().map(str::as_bytes),
+        }
+    }
+}
+
+#[self_referencing]
+pub struct RopeOrSliceRopeTextProviderIterator<'a> {
+    rope_slice: RopeSlice<'a>,
+
+    #[borrows(rope_slice)]
+    chunks_iterator: Chunks<'a>,
+}
+
+impl<'a> Iterator for RopeOrSliceRopeTextProviderIterator<'a> {
+    type Item = &'a str;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        self.with_chunks_iterator_mut(|chunks_iterator| chunks_iterator.next())
+    }
+}
+
+// I believe this type can't be Copy/Clone in order for the
+// `get_captures()` unsafe stuff to be sound
+pub struct CaptureInfo<'a> {
+    pub node: Node<'a>,
+    pub pattern_index: usize,
+}
+
+#[self_referencing]
+pub struct Captures<'a, 'text: 'a, 'tree: 'a> {
+    text: RopeOrSlice<'text>,
+    query_cursor: QueryCursor,
+    query: &'a Query,
+    filter: Option<&'a Filterer>,
+    tree: Cow<'tree, Tree>,
+    capture_index: u32,
+    #[borrows(text, mut query_cursor, query, tree)]
+    #[covariant]
+    captures_iterator: QueryCaptures<'this, 'this, 'this, RopeOrSlice<'this>>,
```

Comment:
I initially tried to "go ahead" and call the `.filter_map()` (that in the previous versions of this pattern was always getting called to narrow the returned captures) on the `QueryCursor::captures()` iterator (converted to a `StreamingIterator`) but ran into "lifetime stuff" including some stuff around covariant/invariant "subtyping"

But so keeping this field "simpler" and then basically "manually" implementing `StreamingIterator` (including "inlining" the `.filter_map()` logic) made it more straightforward